src/node/node.rs

```diff
@@ -0,0 +1,951 @@
+                );
+
+                self._broadcast_pbft_message(
+                    pbft_message.get_info().get_seq_num(),
+                    PbftMessageType::Prepare,
+                    (*pbft_message.get_block()).clone(),
+                )?;
+            }
+
+            PbftMessageType::Prepare => {
+                let pbft_message = protobuf::parse_from_bytes::<PbftMessage>(&msg.content)
+                    .map_err(|e| PbftError::SerializationError(e))?;
+
+                self._handle_not_ready(multicast_not_ready, &pbft_message, msg.content.clone())?;
+
+                self.msg_log.add_message(pbft_message.clone());
+
+                self.msg_log.prepared(&pbft_message, self.state.f)?;
+
+                if self.state.phase != PbftPhase::Checking {
+                    self.state.switch_phase(PbftPhase::Checking);
+                    info!("{}: Checking blocks", self.state);
+                    self.service
+                        .check_blocks(vec![
+                            BlockId::from(pbft_message.get_block().clone().block_id),
+                        ])
+                        .expect("Failed to check blocks");
+                }
+            }
+
+            PbftMessageType::Commit => {
+                let pbft_message = protobuf::parse_from_bytes::<PbftMessage>(&msg.content)
+                    .map_err(|e| PbftError::SerializationError(e))?;
+
+                self._handle_not_ready(multicast_not_ready, &pbft_message, msg.content.clone())?;
+
+                self.msg_log.add_message(pbft_message.clone());
+
+                self.msg_log.committed(&pbft_message, self.state.f)?;
+
+                self.state.switch_phase(PbftPhase::FinalCommitting);
+
+                self._broadcast_pbft_message(
+                    pbft_message.get_info().get_seq_num(),
+                    PbftMessageType::CommitFinal,
+                    (*pbft_message.get_block()).clone(),
+                )?;
+            }
+
+            PbftMessageType::CommitFinal => {
+                let pbft_message = protobuf::parse_from_bytes::<PbftMessage>(&msg.content)
+                    .map_err(|e| PbftError::SerializationError(e))?;
+
+                self._handle_not_ready(multicast_not_ready, &pbft_message, msg.content.clone())?;
+
+                self.msg_log.add_message(pbft_message.clone());
+
+                self.msg_log.check_msg_against_log(&&pbft_message, true, self.state.f + 1)?;
+
+                if self.state.phase == PbftPhase::FinalCommitting {
+                    let working_block = if let WorkingBlockOption::WorkingBlock(ref wb) = self.state.working_block {
+                        Ok(wb.clone())
+                    } else {
+                        Err(PbftError::NoWorkingBlock)
+                    }?;
+
+                    self.state.switch_phase(PbftPhase::Finished);
+
+                    // Don't commit if we've seen this block already, but go ahead if we somehow
```

Comment:
Would skipping a block be possible? As in catchup?

Reply:
In my testing, I encountered a few sparse instances where the timing of the system was such that blocks were delivered out of order